### ISSUE TITLE
feign with hystrix

### ISSUE THREAD

**Issue by jdamick** [Netflix/feign#189]
I'd like to use hystrix with feign, but it appears i'd have to shoehorn it into the invocation handler factory, and if i wanted the ability to do things async it would be more changes to methodmetadata to look for Future<?> return types..    
I could do this on the outside of the feign, but that seems kludgy so i wanted to see if there were any better ideas floating around?

**Comment by adriancole** [comment]
We could start (like beta) with invocation handler.

I suspect any async stuff (rx or hystrix) would imply contract extensions
of some sort and possibly also deeper looks or refactoring around
invocation handler.

I had considered a decorator feign when thinking about rxjava. Ex

new HystrixFeign.Builder () which intercepts or weaves things properly so
as to make tricky things easier and also availing the integration to the
feign package to access what it needs (either by package accessor or
literally being in feign package). With a builder like this, we could make
any details like invocationhandler implicit and also decide which contracts
to support, in case we haven't implemented extensions yet.

Make sense?

**Comment by jdamick** [comment]
Yes, that's similar to what i was thinking.  That would make rxjava and/or hystrix or whatever else a lot easier to add as a bolt-on. I wonder if makes sense to add a builder "interceptor", something analogous to the requestInceptors, that would let you just pluign into the Feign.Builder but keep that same style of registering add-ons?

**Comment by adriancole** [comment]
Which bolt-ons do you think you would use together with the same feign
instance? One concern about bolt-on builder interceptor is whether or not
bolt-ons are compatible. Ie it is easier to support hystrix or rxjava vs
both

**Comment by spencergibb** [comment]
I had done some work on this a while ago [here: HystrixInvocationHandler](https://github.com/spencergibb/halfpipe/blob/master/halfpipe-client/src/main/java/halfpipe/client/HystrixInvocationHandler.java).  It required some funkyness in a [decoder](https://github.com/spencergibb/halfpipe/blob/master/halfpipe-client/src/main/java/halfpipe/client/HystrixJacksonDecoder.java).  I'm sure something cleaner can be worked out.

**Comment by adriancole** [comment]
Thanks for the tips, Spencer!

**Comment by rubesMN** [comment]
Are there any updates to the progress on this?   Would be awesome to combine Hystrix annotations with feign's integrations with ribbon into spring cloud.   How does 9.0.0 fit into Spring Cloud's labeling/releases?

**Comment by spencergibb** [comment]
When stable releases of Netflix tools come out, we can integrate them into Spring Cloud.  There's no timeline.

**Comment by rubesMN** [comment]
Perhaps I'm confused...  I'm actually modifying an example you wrote called feign-eureka,.. and feign, a netflix tool I thought, is being used.  Hystrix, has a org.springframework.cloud artifact called spring-cloud-starter-hystrix.  Both feign and hystrix are part of the dependencies I get when I use a parent in my pom of 'spring-cloud-starter-parent'.   Was this question specific to netflix's rxjava toolset?

**Comment by spencergibb** [comment]
@rubesMN you asked about spring cloud in a netflix issue.  Spring Cloud Netflix's issues are [here](https://github.com/spring-cloud/spring-cloud-config/issues).  @adriancole has no visibility into Spring Cloud Netflix, just Feign.  I just happen to lurk around netflix projects :-)

**Comment by adriancole** [comment]
@rubesMN so I planned to put more time into Feign after cutting denominator 4.5 which happened yesterday. So.. timeline is uhh weeks? Definitely analysis will happen in the next couple weeks.

**Comment by adriancole** [comment]
@rubesMN, @spencergibb I'm going to delete the conversation about spring cloud releases etc. I want to start this in earnest and not have folks distracted. Hope that's ok! (ps. will delete this message too!)

**Comment by adriancole** [comment]
OK folks. Let's please enumerate the features we want from Hystrix? We also need to agree on the scope that Hystrix integration plugs into.

Seems the most sensible scope is `feign.Target` as that corresponds with a seam in a SOA more than an individual http request would. Ex. circuit-breaking might fit here? Features such as fallback likely have finer-grained scope.

In order to design this, we really need a story of what's desired. cc @mattrjacobs

**Comment by spencergibb** [comment]
Maybe start with something like [ribbon's hystrix annotation support](https://github.com/Netflix/ribbon/tree/master/ribbon)?  For me configuration via Archaius or annotations (ala [javanica](https://github.com/Netflix/Hystrix/tree/master/hystrix-contrib/hystrix-javanica/)) is important.  Also access to the Rx and Async results is very important.

**Comment by mattrjacobs** [comment]
@adriancole Happy to take a look at whatever ideas you've got.  Is there anything concrete I can provide at this point?

**Comment by jdamick** [comment]
Right now I just need basic circuit breaking - open/closed, fallback is bonus credit, and it would be fine to have it apply to the whole client rather than per request as you mention.  Also queued vs. non-queued invocation per method and for me it's fine if it's simply one or the other and not both per method.

**Comment by adriancole** [comment]
Here's some straw-man ideas for what seems to be easiest.
## Implicit hystrixification
We implicitly "hystrix command" existing code. This allows folks to plug-in and magically get hystrixified.
```java
// this is implicitly hystrixed and calls .execute()
List<Contributors> list();
// if you want async, you need to go via hystrix
HystrixCommand<List<Bucket>> list();
```
## Only expose HystrixCommand
Instead of exposing Future or Observable directly, we expose HystrixCommand which can give you one or the other. That's mainly because we don't currently have plans for another Future implementation. This also makes interactions with other components precise. Ex. we expect no other integration to claim the type `HystrixCommand`.
```java
// if you want async, you need to go via hystrix, as we won't magically cover types derived from HystrixCommand
HystrixCommand<List<Contributors>> list();
```
## HystrixCommand name/group mapping
There are two main things we have to work with `Target.name()` and `Feign.methodKey`. 

 * `Target.name()` would differentiate multiple api of the same type, typically by endpoint. For example, this is the ribbon client name.
 * `MethodMetadata.configKey()` is a pre-parsed javadoc version of the java method invoked. Ex. `S3#listBuckets()` or `S3#listBucketsByName(String)` 

So, if we have a metadata that can say "S3 on this endpoint" (aka targetName) and "this precise rest call" (aka configKey)

### HystrixCommand group mapping.. can it be targetName?
Is targetName aka ribbon client name, or some user-provided name a good Hystrix group?

### HystrixCommand name mapping.. can it be configKey? can it be parsed only once?
Bear in mind that in many cases method overloads will need the same behavior. Ex. `S3#listBuckets()` or `S3#listBucketsByName(String)` might have the same fallback, and probably the same timeouts?

Should we use configKeys at all? Should our hystrix integration include a command name mapper that's parses configKey and returns a hystrix command name? Could such a thing be done once?

@mattrjacobs thoughts?

**Comment by adriancole** [comment]
@spencergibb fyi, I'm intentionally leaving additional annotation processing for later :) Not saying we won't, just I'd like to solve the basics. For example, similar to passing type adapters to `GsonDecoder`, our first version of hystrix support could allow folks to pass what they'd otherwise do via annotation processing. That decouples this from the separate challenge of coordinated plugins (ex. jaxrs + hystrix vs default + ribbon + hystrix)

**Comment by spencergibb** [comment]
@adriancole np.  Solving for basics first is a great way to go.  In your strawman, you only touch lightly on fallback.  In javanica, fallback declaration is a method name on an annotation, in Ribbon 2 it is a class reference on an annotation.  Any ideas on how we might declare fallback?

**Comment by adriancole** [comment]
@spencergibb I agree we should handle fallbacks and ideally at
declaration time. Maybe we can straw-man just using javanica's
annotation if present. I want to avoid annotation if statements though
:) basically, I don't want folks to expect us to parse this and also
ribbon etc, if that makes sense.
https://github.com/Netflix/Hystrix/blob/master/hystrix-contrib/hystrix-javanica/src/main/java/com/netflix/hystrix/contrib/javanica/annotation/HystrixCommand.java

**Comment by adriancole** [comment]
personally, I'm a week deferred due to unrelated work.

Next step is to get some sort of consensus or at least opinion on HystrixCommand name mapping

**Comment by mattrjacobs** [comment]
Sorry for the delay in responding.  I'm not an expert in Feign concepts, but it looks like you've hit on a lot of the interesting seams.

As @adriancole mentioned above, the granularities of different Hystrix objects are important.  Here are some concepts, some of which you might want, some of which you might not.

* Command - Single-use object that wraps a single invocation
* Command Key - Per-JVM object that is the canonical was to distinguish classes of Commands
* Command Properties - Per-JVM object that stores configuration info for all Commands that share a Command Key (things like timeout value/fallback enabled/circuit forced open/etc)
* Command Metrics - Per-JVM object that stores runtime behavior for all Commands that share a Command Key (things like countSuccess/countTimeout/countFallbackSuccess/etc)

Per Command Key, there should be a single Java class that implements a `run()` method, and optionally a `getFallback()` method.  This behavior (along with config and metrics) gets stitched together by code in Hystrix.

* Thread Pool - Per-JVM thread pool that multiple Commands (of any Command Key) can run in
* Thread Pool Key - Per-JVM object that is the canonical way to distinguish thread pools
* Thread Pool Properties/Metrics - similar to above

* Group - Per-JVM object that aggregates commands (regardless of Thread Pool Key)
* Group Key - Per-JVM object that is the canonical way to distinguish groups

So some of the pieces above are perfectly reasonable to set at the method invocation level, some (like thread pool size) are necessarily shared between HystrixCommands and are harder to represent in this fashion.

**Comment by adriancole** [comment]
@mattrjacobs fantastic info. Thanks so much for taking time to share this.

So, I think I've enough information to start an impl. One clarification:

Say we have a Command Key for cloudwatch put-metric-data (hypothetically). Is there a hystrix notion to express the difference between this pointed to us-east-1 vs us-west-2?

**Comment by adriancole** [comment]
another issue related to fallback values (I think) https://github.com/Netflix/feign/issues/249 from @clatko. Maybe we should open an umbrella issue for fallback related things, so that they can be collected more coherently than hystrix sidebars..
